File: src/drivers/vga.rs
use core::fmt;
use volatile::Volatile;
use spin::Mutex;
use lazy_static::lazy_static;

const VGA_WIDTH: usize = 80;
const VGA_HEIGHT: usize = 25;
const VGA_BUFFER_ADDR: usize = 0xB8000;

#[repr(u8)]
#[allow(dead_code)]
pub enum Color {
    Black = 0, Blue = 1, Green = 2, Cyan = 3, Red = 4,
    Magenta = 5, Brown = 6, LightGray = 7, DarkGray = 8,
    LightBlue = 9, LightGreen = 10, LightCyan = 11, LightRed = 12,
    Pink = 13, Yellow = 14, White = 15,
}

#[derive(Clone, Copy)]
#[repr(transparent)]
struct ColorCode(u8);

impl ColorCode {
    fn new(fg: Color, bg: Color) -> ColorCode {
        ColorCode((bg as u8) << 4 | (fg as u8))
    }
}

#[derive(Clone, Copy)]
#[repr(C)]
struct ScreenChar {
    ascii_char: u8,
    color_code: ColorCode,
}

#[repr(transparent)]
struct Buffer {
    chars: [[Volatile<ScreenChar>; VGA_WIDTH]; VGA_HEIGHT],
}

pub struct Writer {
    row: usize,
    col: usize,
    color_code: ColorCode,
    buffer: &'static mut Buffer,
}

impl Writer {
    pub fn write_byte(&mut self, byte: u8) {
        match byte {
            b'\n' => self.new_line(),
            0x08 => self.backspace(), // backspace
            byte => {
                if self.col >= VGA_WIDTH {
                    self.new_line();
                }
                let row = self.row;
                let col = self.col;
                self.buffer.chars[row][col].write(ScreenChar {
                    ascii_char: byte,
                    color_code: self.color_code,
                });
                self.col += 1;
            }
        }
    }

    pub fn write_string(&mut self, s: &str) {
        for byte in s.bytes() {
            match byte {
                0x20..=0x7e | b'\n' | 0x08 => self.write_byte(byte),
                _ => self.write_byte(0xfe), // unprintable -> block char
            }
        }
    }

    fn backspace(&mut self) {
        if self.col > 0 {
            self.col -= 1;
            let row = self.row;
            let col = self.col;
            self.buffer.chars[row][col].write(ScreenChar {
                ascii_char: b' ',
                color_code: self.color_code,
            });
        }
    }

    fn new_line(&mut self) {
        if self.row < VGA_HEIGHT - 1 {
            self.row += 1;
        } else {
            for row in 1..VGA_HEIGHT {
                for col in 0..VGA_WIDTH {
                    let ch = self.buffer.chars[row][col].read();
                    self.buffer.chars[row - 1][col].write(ch);
                }
            }
            self.clear_row(VGA_HEIGHT - 1);
        }
        self.col = 0;
    }

    fn clear_row(&mut self, row: usize) {
        let blank = ScreenChar { ascii_char: b' ', color_code: self.color_code };
        for col in 0..VGA_WIDTH {
            self.buffer.chars[row][col].write(blank);
        }
    }

    pub fn clear_screen(&mut self) {
        for row in 0..VGA_HEIGHT {
            self.clear_row(row);
        }
        self.row = 0;
        self.col = 0;
    }
}

impl fmt::Write for Writer {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        self.write_string(s);
        Ok(())
    }
}

lazy_static! {
    pub static ref WRITER: Mutex<Writer> = Mutex::new(Writer {
        row: 0,
        col: 0,
        color_code: ColorCode::new(Color::LightGray, Color::Black),
        buffer: unsafe { &mut *(VGA_BUFFER_ADDR as *mut Buffer) },
    });
}

pub fn init() {
    WRITER.lock().clear_screen();
}

#[macro_export]
macro_rules! print {
    ($($arg:tt)*) => ($crate::drivers::vga::_print(format_args!($($arg)*)));
}

#[macro_export]
macro_rules! println {
    () => ($crate::print!("\n"));
    ($($arg:tt)*) => ($crate::print!("{}\n", format_args!($($arg)*)));
}

#[doc(hidden)]
pub fn _print(args: fmt::Arguments) {
    use fmt::Write;
    WRITER.lock().write_fmt(args).unwrap();
}

File: src/drivers/vga.cpp
#include "vga.hpp"

namespace VGA {

Writer::Writer()
    : color_code_(static_cast<std::uint8_t>(Color::LightGray)),
      buffer_(reinterpret_cast<volatile ScreenChar*>(BUFFER_ADDR)) {}

void Writer::clear_row(std::size_t row) {
    ScreenChar blank{' ', color_code_};
    for (std::size_t col = 0; col < WIDTH; ++col) {
        buffer_[row * WIDTH + col] = blank;
    }
}

void Writer::clear_screen() {
    for (std::size_t row = 0; row < HEIGHT; ++row) clear_row(row);
    row_ = 0;
    col_ = 0;
}

void Writer::new_line() {
    if (row_ < HEIGHT - 1) {
        ++row_;
    } else {
        for (std::size_t row = 1; row < HEIGHT; ++row) {
            for (std::size_t col = 0; col < WIDTH; ++col) {
                buffer_[(row - 1) * WIDTH + col] = buffer_[row * WIDTH + col];
            }
        }
        clear_row(HEIGHT - 1);
    }
    col_ = 0;
}

void Writer::write_byte(char c) {
    if (c == '\n') { new_line(); return; }
    if (col_ >= WIDTH) new_line();
    buffer_[row_ * WIDTH + col_] = ScreenChar{static_cast<std::uint8_t>(c), color_code_};
    ++col_;
}

void Writer::write_string(const char* s) {
    while (*s) write_byte(*s++);
}

Writer writer;

void init() { writer.clear_screen(); }
void print(const char* s) { writer.write_string(s); }
void println(const char* s) { writer.write_string(s); writer.write_byte('\n'); }

void write_uint_padded(unsigned value, int width) {
    char buf[16];
    int i = 0;
    if (value == 0) buf[i++] = '0';
    while (value > 0) { buf[i++] = static_cast<char>('0' + (value % 10)); value /= 10; }
    for (int pad = i; pad < width; ++pad) writer.write_byte('0');
    while (i > 0) writer.write_byte(buf[--i]);
}

} // namespace VGA